client/asset/btc/btc.go

```diff
@@ -1136,132 +1153,320 @@ func (btc *ExchangeWallet) LocktimeExpired(contract dex.Bytes) (bool, time.Time,
 		if err != nil {
-			return nil, fmt.Errorf("spending transaction not found")
+			btc.log.Debugf("FindRedemption - ended at block %d", blk.Height)
+			// TODO: This assumes the block does not exist but could it be a different error?
+			return nil
 		}
+		nextBlockHash = nbh.String()
+	}
+}
+
+// foundRedemption returns the specified contract's secret and the txid, input
+// id of the spending input via the contract's registered result channel. Also
+// purges the contract from the findRedemptionQueue.
+// This method MUST be called with the findRedemptionMtx lock held for writes.
+func (dcr *ExchangeWallet) foundRedemption(foundIn string, contractOp outPoint, redeemTxId string, inputIndex int, secret []byte) {
+	redeemTxHash, err := chainhash.NewHashFromStr(redeemTxId)
+	if err != nil {
+		dcr.findRedemptionError(contractOp, foundIn, err, true)
+		return
+	}
+
+	req, exists := dcr.findRedemptionQueue[contractOp]
+	delete(dcr.findRedemptionQueue, contractOp)
+	if !exists {
+		dcr.log.Debugf("missing request for found redemption (%s, found in %s)", contractOp.String(), foundIn)
+		return
 	}
-	// No viable way to get here.
+
+	dcr.log.Debugf("redemption for contract %s found in %s", contractOp.String(), foundIn)
+	req.resultChan <- &asset.FindRedemptionResult{
+		ContractCoinID:   toCoinID(&contractOp.txHash, contractOp.vout),
+		RedemptionCoinID: toCoinID(redeemTxHash, uint32(inputIndex)),
+		Secret:           secret,
+	}
+	close(req.resultChan)
+}
+
+// findRedemptionError returns any error encountered while attempting to find
+// redemption info for the specified contract via the contract's registered
+// result channel.
+// If purge==true, the contract is deleted from the findRedemptionQueue.
+// This method MUST be called with the findRedemptionMtx lock held for writes.
+func (btc *ExchangeWallet) findRedemptionError(contractOp outPoint, lastScanned string, err error, purge bool) {
+	req, exists := btc.findRedemptionQueue[contractOp]
+	if purge {
+		delete(btc.findRedemptionQueue, contractOp)
+	}
+	if !exists {
+		btc.log.Debugf("missing request for find redemption error (%s, %s): %v", contractOp.String(), lastScanned, err)
+		return
+	}
+
+	btc.log.Debugf("error finding redemption for %s in %s: %v", contractOp.String(), lastScanned, err)
+	req.resultChan <- &asset.FindRedemptionResult{
+		ContractCoinID: toCoinID(&contractOp.txHash, contractOp.vout),
+		Err:            err,
+	}
+	close(req.resultChan)
+}
+
+// tryExtractContractSecret checks if the passed tx input spends the specified
+// contract. If so, the contract is removed from the findRedemptionQueue. The
+// contract's secret or any error encountered while trying to parse the secret
+// is propagated via the registered result channel.
+func (btc *ExchangeWallet) tryExtractContractSecret(input btcjson.Vin, contractOp outPoint, contractHash []byte) ([]byte, error) {
+	if input.Vout != contractOp.vout || input.Txid != contractOp.txHash.String() {
+		return nil, nil
```

Comment:
It'd be difficult to later tell if an error returned by this method is a secret parse error or indicative of non-matching prev out. Also, the method is prefixed `tryExtract` which means it should only error if it tried to extract but failed to; if it doesn't even try, shouldn't be an error.

Didn't worry too much about the intuitiveness because I've seen similar code anyway, will call this tidbit out in the method's doc so callers are better informed. If others think it's more intuitive to return an error here, I'll reconsider. Cool?

Reply:
You can give errors types, so one error has a different meaning that the other and can be handled differently. Yes it's more code, but as I commented on `setBestBlock` it could lead to bugs when/if this code ever changes, even if documented.

But I would like to hear what the others think. Maybe I'm overreacting.

Reply:
Update, after reading your later comment relating to this, I've decided to retire this method, it's just a few lines after all. I think having these lines in the tx inputs loop reads better too.